r/R/data.R

```diff
@@ -6,6 +6,214 @@
 # option. This file may not be copied, modified, or distributed
 # except according to those terms.
 
+#' Create a small demo matrix and fragment object.
+#' 
+#' Downloads a 10x Genomics dataset consisting of 3k cells, then performs optional QC and subsetting.  Holds subsetted objects in disk,
+#' and returns a list with both the matrix and fragments. 
+#' @param directory (character) Where the input/output data should be stored.  If NULL, a temporary directory is created.
+#' @param filter_qc (bool) Whether to filter both the RNA and ATAC data using QC information.
+#' @param subset (bool) Whether to subset to only genes/insertions on chromosome 4 and 11.
+#' @param timeout (numeric) Timeout for downloading files in seconds.
+#' @param remove_input_data (logical) Whether to remove the downloaded non-procesed matrix, frags, gencode transcripts, and gencode genes
+#' after processing.
+#' @return (list) A list with the RNA matrix under the name "mat", and the ATAC fragments under the name "frags".
+#' @details
+#' This function downloads the 10x Genomics PBMC 3k dataset. 
+#' Filtering using QC information on the fragments and matrix provides cells with at least 1000 reads, 1000 frags, and a minimum tss enrichment of 10.
+#' Subsetting provides only genes and insertions on chromosomes 4 and 11.
+#' The name of the matrix and fragments folders are `demo_mat` and `demo_frags` respectively.  
+#' Additionally, choosing to qc filter appends a `_filtered`, and choosing to subset data appends a `_subsetted` to the name.
+#' @keywords internal
+prepare_demo_data <- function(directory = NULL, filter_qc = TRUE, subset = TRUE, timeout = 300, remove_input_data = TRUE) {
+    if (is.null(directory)) {
+      directory <- file.path(tempdir())
+      dir.create(directory, recursive = TRUE, showWarnings = FALSE)
+    }
```

Comment:
It might be nicer to create a subdir just for the downloaded files, then add an `on.exit()` to delete the directory so it's more foolproof to clean up downloads